types/alpinejs/index.d.ts

```diff
@@ -0,0 +1,225 @@
+// Type definitions for alpinejs 3.7
+// Project: https://github.com/alpinejs/alpine
+// Definitions by: Thomas Wirth <https://github.com/wtho>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import { ReactiveEffect, UnwrapNestedRefs, ReactiveEffectOptions, reactive as VueReactive } from '@vue/reactivity';
+
+interface Evaluator {
+    (el: Node, expression: string /*, extras?: unknown */): (resultCallback: (result: unknown) => void) => void;
+}
+
+interface ReactiveEffectRunner<T = any> {
+    (): T;
+    effect: ReactiveEffect;
+}
+
+interface ReactivityEngine {
+    reactive: Alpine['reactive'];
+    release: Alpine['release'];
+    effect: Alpine['effect'];
+    raw: Alpine['raw'];
+}
+
+type Walker = (el: Node, callback: (el: Node, skip: () => void) => void) => void;
+export type AttributeTransformer<T = unknown> = (args: { name: string; value: T }) => { name: string; value: T };
+
+interface XDataContext {
+    /**
+     * Will be executed before Alpine initializes teh rest of the component.
+     */
+    init?(): void;
+    [stateKey: string]: any;
+}
+type XData = XDataContext | string | number | boolean;
+
+interface DirectiveUtilities {
+    Alpine: Alpine;
+    effect: () => void;
+    cleanup: () => void;
+    evaluate: (expression: string) => unknown;
+    evaluateLater: (expression: string) => (result: unknown) => void;
+}
+interface DirectiveParameters {
+    value: string;
+    modifiers: string[];
+    expression: string;
+    original: string;
+    type: string;
+}
+
+export interface Alpine {
+    // following TSDoc under MIT was taken from
+    // https://github.com/vuejs/vue-next/blob/92f11d6740929f5b591740e30ae5fba50940ec82/packages/reactivity/src/reactive.ts#L65-L87
+    /**
+     * Creates a reactive copy of the original object.
+     * The reactive conversion is "deep"—it affects all nested properties. In the
+     * ES2015 Proxy based implementation, the returned proxy is **not** equal to the
+     * original object. It is recommended to work exclusively with the reactive
+     * proxy and avoid relying on the original object.
+     *
+     * A reactive object also automatically unwraps refs contained in it, so you
+     * don't need to use `.value` when accessing and mutating their value:
+     *
+     * ```js
+     * const count = ref(0)
+     * const obj = reactive({
+     *   count
+     * })
+     *
+     * obj.count++
+     * obj.count // -> 1
+     * count.value // -> 1
+     * ```
+     */
+    reactive<T extends object>(target: T): UnwrapNestedRefs<T>;
```

Comment:
Thanks for your feedback!

I disagree on this. Because a property is implemented with a getter, it does not mean we have to use `unknown` as the type. If a getter returns a `string`, we would not want to type it as `myGetter: unknown`, but instead `myGetter: string` to indicate the type of the returned value.

The `reactive` getter returns a function of the type `<T extends object>(target: T) => UnwrapNestedRefs<T>`, so we could type the `Alpine.reactive` like one of these:
```ts
reactive: <T extends object>(target: T) => UnwrapNestedRefs<T>;
reactive<T extends object>(target: T): UnwrapNestedRefs<T>;
```

I chose the second version, if you prefer the first one I am open to it, but replacing the type with `unknown` seems not to be a good option here.

(same applies to `release`, `effect` and `raw` below)

Reply:
I don't see any parameter in the source code so I don't understand why it is there. The getter shouldn't have parameters.

Reply:
`reactive` in [`alpine.js`](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/alpine.js#L20) is imported from `./reactivity.js` in [line 1](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/alpine.js#L1).
There in `reactivity.js`, `reactive` is declared in [line 4](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/reactivity.js#L4), set in [line 16](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/reactivity.js#L16) using `setReactivityEngine()` and exported in [line 59](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/reactivity.js#L59).

The reactivity engine is at Alpine.js' core and will be configured early during framework initialization.

`setReactiveEngine` is only called in two places:
* [`alpine/src/index.js`](https://github.com/alpinejs/alpine/blob/63a520d052597b46c7ede82f32d774106ab8fff4/packages/alpinejs/src/index.js#L39-L41) when running Alpine.js in default mode
* [`csp/src/index.js`](https://github.com/alpinejs/alpine/blob/63a520d052597b46c7ede82f32d774106ab8fff4/packages/csp/src/index.js#L5-L6) when running Alpine.js in CSP mode

In both cases, `reactive` is the function of `@vue/reactivity`'s `reactive`, so in my opinion, types which correspond to this function, should be applied.

(same applies to `release`, `effect` and `raw`)

Reply:
@martin-badin can you agree with my argumentation?

Another indicator why this is the correct typing (for `reactive` and `effect` in particular) is the official Alpine.js documentation:

>  Alpine uses VueJS's reactivity engine under the hood to provide these functions. → Read more about @vue/reactivity

> Let's first look at Alpine.reactive(). This function accepts a JavaScript object as its parameter and returns a "reactive" version of that object. For example:
> ```js
> let data = { count: 1 }
> 
> let reactiveData = Alpine.reactive(data)
> ```
> Under the hood, when Alpine.reactive receives data, it wraps it inside a custom JavaScript proxy.
> 
> A proxy is a special kind of object in JavaScript that can intercept "get" and "set" calls to a JavaScript object.


See https://alpinejs.dev/advanced/reactivity

Reply:
@wtho thank you for your explanation. I can agree with you.